File: backend/src/graphql/types/topic.rs
use async_graphql::SimpleObject;
use chrono::{DateTime, Utc};
use uuid::Uuid;

use crate::storage::topics::TopicRow;

#[derive(SimpleObject, Clone)]
pub struct Topic {
    pub id: Uuid,
    pub session_id: Uuid,
    pub title: String,
    pub description: Option<String>,
    pub order_index: i32,
    pub is_completed: bool,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

impl From<TopicRow> for Topic {
    fn from(row: TopicRow) -> Self {
        Self {
            id: row.id,
            session_id: row.session_id,
            title: row.title,
            description: row.description,
            order_index: row.order_index,
            is_completed: row.is_completed,
            created_at: row.created_at,
            updated_at: row.updated_at,
        }
    }
}

File: backend-python/app/models/topic.py
import uuid
from datetime import datetime

from sqlalchemy import (
    Boolean,
    DateTime,
    ForeignKey,
    Index,
    Integer,
    String,
    func,
)
from sqlalchemy.dialects.postgresql import ARRAY, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class Topic(Base):
    __tablename__ = "topics"
    __table_args__ = (Index("idx_topics_session", "session_id"),)

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    session_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("study_sessions.id", ondelete="CASCADE"),
        nullable=False,
    )
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    subtopics: Mapped[list[str]] = mapped_column(
        ARRAY(String), nullable=False, server_default="{}"
    )
    order_index: Mapped[int] = mapped_column(Integer, nullable=False)
    is_completed: Mapped[bool] = mapped_column(
        Boolean, nullable=False, server_default="false"
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    session: Mapped["StudySession"] = relationship(  # noqa: F821
        back_populates="topics"
    )
    chat: Mapped["Chat | None"] = relationship(  # noqa: F821
        back_populates="topic", cascade="all, delete-orphan"
    )
